rlwe/keys.go

```diff
@@ -10,213 +17,722 @@ type SecretKey struct {
+//
+// Unless w implements the buffer.Writer interface (see lattigo/utils/buffer/writer.go),
+// it will be wrapped into a bufio.Writer. Since this requires allocations, it
+// is preferable to pass a buffer.Writer directly:
+//
+//   - When writing multiple times to a io.Writer, it is preferable to first wrap the
+//     io.Writer in a pre-allocated bufio.Writer.
+//   - When writing to a pre-allocated var b []byte, it is preferable to pass
+//     buffer.NewBuffer(b) as w (see lattigo/utils/buffer/buffer.go).
+func (sk SecretKey) WriteTo(w io.Writer) (n int64, err error) {
+	return sk.Value.WriteTo(w)
 }
 
-// LevelP returns the level of the modulus P of the target.
-// Returns -1 if P is absent.
-func (sk *SecretKey) LevelP() int {
-	if sk.Value.P != nil {
-		return sk.Value.P.Level()
-	}
-
-	return -1
+// ReadFrom reads on the object from an io.Writer. It implements the
+// io.ReaderFrom interface.
+//
+// Unless r implements the buffer.Reader interface (see see lattigo/utils/buffer/reader.go),
+// it will be wrapped into a bufio.Reader. Since this requires allocation, it
+// is preferable to pass a buffer.Reader directly:
+//
+//   - When reading multiple values from a io.Reader, it is preferable to first
+//     first wrap io.Reader in a pre-allocated bufio.Reader.
+//   - When reading from a var b []byte, it is preferable to pass a buffer.NewBuffer(b)
+//     as w (see lattigo/utils/buffer/buffer.go).
+func (sk *SecretKey) ReadFrom(r io.Reader) (n int64, err error) {
+	return sk.Value.ReadFrom(r)
 }
 
-// NewPublicKey returns a new PublicKey with zero values.
-func NewPublicKey(params Parameters) (pk *PublicKey) {
-	return &PublicKey{CiphertextQP{Value: [2]ringqp.Poly{params.RingQP().NewPoly(), params.RingQP().NewPoly()}, MetaData: MetaData{IsNTT: true, IsMontgomery: true}}}
+// MarshalBinary encodes the object into a binary form on a newly allocated slice of bytes.
+func (sk SecretKey) MarshalBinary() (p []byte, err error) {
+	return sk.Value.MarshalBinary()
 }
 
-// LevelQ returns the level of the modulus Q of the target.
-func (pk *PublicKey) LevelQ() int {
-	return pk.Value[0].Q.Level()
+// UnmarshalBinary decodes a slice of bytes generated by
+// MarshalBinary or WriteTo on the object.
+func (sk *SecretKey) UnmarshalBinary(p []byte) (err error) {
+	return sk.Value.UnmarshalBinary(p)
 }
 
-// LevelP returns the level of the modulus P of the target.
-// Returns -1 if P is absent.
-func (pk *PublicKey) LevelP() int {
-	if pk.Value[0].P != nil {
-		return pk.Value[0].P.Level()
+func (sk *SecretKey) isEncryptionKey() {}
+
+type VectorQP []ringqp.Poly
+
+// NewVectorQP returns a new PublicKey with zero values.
+func NewVectorQP(params ParameterProvider, size, levelQ, levelP int) (v VectorQP) {
+	rqp := params.GetRLWEParameters().RingQP().AtLevel(levelQ, levelP)
+
+	v = make(VectorQP, size)
+
+	for i := range v {
+		v[i] = rqp.NewPoly()
 	}
 
-	return -1
+	return
+}
+
+func (p VectorQP) LevelQ() int {
+	return p[0].LevelQ()
```

Comment:
Is it safe that p.length != 0?
Are those function used only for tests?

Reply:
Done & updated godoc